src/Forms/SearchableDropdownTrait.php

```diff
@@ -0,0 +1,594 @@
+    public function Field($properties = []): DBHTMLText
+    {
+        // The entwine class needs to be added so that it is rendered in an entwine context
+        $this->addExtraClass('entwine');
+        return $this->customise($properties)->renderWith(self::class);
+    }
+
+    public function saveInto(DataObjectInterface $record): void
+    {
+        if (empty($record)) {
+            return;
+        }
+        $name = $this->getName();
+        $ids = $this->getValueArray();
+        if (substr($name, -2) === 'ID') {
+            // has_one field
+            $record->$name = $ids[0] ?? 0;
+            $record->write();
+        } else {
+            // has_many / many_many field
+            if (!method_exists($record, 'hasMethod')) {
+                throw new LogicException('record does not have method hasMethod()');
+            }
+            /** @var DataObject $record */
+            if (!$record->hasMethod($name)) {
+                throw new LogicException("Relation $name does not exist");
+            }
+            /** @var Relation $relation */
+            $relation = $record->$name();
+            if (!is_a($relation, Relation::class)) {
+                throw new LogicException("$name is not a Relation");
+            }
+            $relation->setByIDList($ids);
+        }
+    }
+
+    /**
+     * @param Validator $validator
+     */
+    public function validate($validator): bool
+    {
+        return $this->extendValidationResult(true, $validator);
+    }
+
+    public function Type(): string
+    {
+        return '';
+    }
+
+    public function getSchemaDataType(): string
+    {
+        if ($this->getIsMultiple()) {
+            return FormField::SCHEMA_DATA_TYPE_MULTISELECT;
+        }
+        return FormField::SCHEMA_DATA_TYPE_SINGLESELECT;
+    }
+
+    /**
+     * Provide data to the JSON schema for the frontend component
+     */
+    public function getSchemaDataDefaults(): array
+    {
+        $data = parent::getSchemaDataDefaults();
+        $data = $this->updateDataForSchema($data);
+        $name = $this->getName();
+        if ($this->getIsMultiple() && strpos($name, '[') === false) {
+            $name .= '[]';
+        }
+        $data['name'] = $name;
+        $data['disabled'] = $this->isDisabled() || $this->isReadonly();
+        if ($this->getIsLazyLoaded()) {
+            $data['optionUrl'] = Controller::join_links($this->Link(), 'search');
+        } else {
+            $data['options'] = array_values($this->getOptionsForSchema()->toNestedArray());
+        }
+        return $data;
+    }
+
+    public function getSchemaStateDefaults(): array
+    {
+        $data = parent::getSchemaStateDefaults();
+        $data = $this->updateDataForSchema($data);
+        return $data;
+    }
+
+    abstract protected function init(): void;
+
+    /**
+     * Get whether the field allows multiple values
+     */
+    protected function getIsMultiple(): bool
+    {
+        return $this->isMultiple;
+    }
```

Comment:
I don't think it's sensible to have a getter here.... it's really only needed for this trait to know whether it's dealing with a `SearchableDropdownField` or `SearchableMultiDropdownField`. I don't think anything other than this trait directly will ever need this method, since any class using it will already know "I want multiple values" or "I only want a single value".

Reply:
Removed